finality-aleph/src/sync/message_limiter.rs

```diff
@@ -32,31 +31,32 @@ impl<'a, D: Encode, const LIMIT: usize> Limiter<'a, D, LIMIT> {
         Self {
             msg,
             start_index: 0,
-            indexes: (0..=msg.len()).collect(),
         }
     }
 
     pub fn next_largest_msg(&mut self) -> Result<Option<&'a [D]>, Error> {
         if self.start_index == self.msg.len() {
             return Ok(None);
         }
-        let end_idx = self.bs_by_encode()?;
+        let end_idx = self.find_idx_of_largest_prefix()?;
 
         let start_index = self.start_index;
         self.start_index = end_idx;
 
         Ok(Some(&self.msg[start_index..end_idx]))
     }
 
-    fn bs_by_encode(&self) -> Result<usize, Error> {
-        let indexes = &self.indexes[self.start_index..];
-
-        let idx = indexes.partition_point(|&idx| {
-            let encoded_size = self.msg[self.start_index..idx].encoded_size();
-            encoded_size <= LIMIT
-        }) - 1; // minus 1 since this is the first index where encoded size is larger than LIMIT
+    fn find_idx_of_largest_prefix(&self) -> Result<usize, Error> {
+        let mut idx = self.start_index;
+        let mut encoded_sum = 0;
+        while idx < self.msg.len() && encoded_sum <= LIMIT {
+            encoded_sum += self.msg[idx].encoded_size();
+            idx += 1;
+        }
 
-        let idx = idx + self.start_index;
+        while idx > self.start_index && self.msg[self.start_index..idx].encoded_size() > LIMIT {
```

Comment:
Add a short comment describing this estimation.

Reply:
done